File: src/app/services/training.service.ts
import { Injectable } from '@angular/core';

import { Decoder, Stream, Utils, Profile } from '@garmin/fitsdk';
import { DurationType, TargetType, Workout, WorkoutStep } from '../model/Workout';
import { MetricService } from './metric.service';
import { ServiceFactory } from './ServiceFactory';
import { Metric, MetricType } from '../model/Metric';


class NewWorkoutEvent extends Event {
  constructor(public workout: Workout) {
    super('newWorkout');
  }
}


@Injectable({
  providedIn: 'root'
})
export class TrainingService {
  private workout: Workout | null = null;
  private eventTarget: EventTarget = new EventTarget();

  private currentStepIndex: number = -1;
  private currentStepStartTime: number = 0;
  private currentStepStartDistance: number = 0;

  private metricService: MetricService;

  constructor() {
    this.metricService = ServiceFactory.getMetricService();
    setInterval(() => {
      this.checkWorkoutUpdate();
    }
      , 500);
  }

  loadFitFile(fileContent: ArrayBuffer): void {
    this.stopWorkout(); // Stop any existing workout before loading a new one

    const stream = Stream.fromByteArray(fileContent);
    const decoder = new Decoder(stream);

    const { messages, errors } = decoder.read();
    if (errors.length > 0) {
      console.error("Errors while reading FIT file:", errors);
    }
    console.log(JSON.stringify(messages, null, 2));
    this.workout = Workout.fromGarminFitFile(messages);
    this.eventTarget.dispatchEvent(new NewWorkoutEvent(this.workout));
  }

  addNewWorkoutListener(callback: (event: NewWorkoutEvent) => void): void {
    this.eventTarget.addEventListener('newWorkout', callback as any);
  }
  removeNewWorkoutListener(callback: (event: NewWorkoutEvent) => void): void {
    this.eventTarget.removeEventListener('newWorkout', callback as any);
  }

  getCurrentWorkout(): Workout | null {
    return this.workout;
  }

  startWorkout(): void {
    if (!this.workout) {
      throw new Error("No workout loaded");
    }
    if (this.currentStepIndex !== -1) {
      throw new Error("Workout already started");
    }
    this.moveToNextStep();
  }

  stopWorkout(): void {
    this.workout = null;
    this.currentStepIndex = -1;
    this.currentStepStartTime = 0;
    this.currentStepStartDistance = 0;
  }

  protected checkWorkoutUpdate(): void {
    if (!this.workout || this.currentStepIndex < 0) {
      return;
    }
    const currentStep = this.currentStep;
    if (!currentStep) {
      return;
    }
    if (this.remainingTime !== null && this.remainingTime <= 0) {
      this.moveToNextStep();
    }
    if (this.remainingDistance !== null && this.remainingDistance <= 0) {
      this.moveToNextStep();
    }
  }

  moveToNextStep(): void {
    if (!this.workout) {
      throw new Error("No workout loaded");
    }
    this.currentStepIndex++;
    this.currentStepStartTime = new Date().getTime();
    this.currentStepStartDistance = this.metricService.getByMetricType(MetricType.DISTANCE)?.getLastValue() || 0;
  }

  moveToPreviousStep(): void {
    if (!this.workout || this.currentStepIndex <= 0) {
      throw new Error("No previous step available");
    }
    this.currentStepIndex--;
    this.currentStepStartTime = new Date().getTime();
    this.currentStepStartDistance = this.metricService.getByMetricType(MetricType.DISTANCE)?.getLastValue() || 0;
  }


  get currentStep(): WorkoutStep | null {
    if (!this.workout || this.currentStepIndex >= this.workout.steps.length || this.currentStepIndex < 0) {
      return null;
    }
    return this.workout.steps[this.currentStepIndex];
  }

  get nextStep(): WorkoutStep | null {
    if (!this.workout || this.currentStepIndex + 1 >= this.workout.steps.length) {
      return null;
    }
    return this.workout.steps[this.currentStepIndex + 1];
  }

  get remainingTime(): number | null {
    if (!this.workout || this.currentStepIndex < 0) {
      return null;
    }
    const currentStep = this.currentStep;
    if (!currentStep || currentStep.durationType !== DurationType.TIME) {
      return null;
    }
    const elapsedTime = (new Date().getTime() - this.currentStepStartTime) / 1000; // Convert to seconds
    return currentStep.durationValue - elapsedTime;
  }

  get remainingDistance(): number | null {
    if (!this.workout || this.currentStepIndex < 0) {
      return null;
    }
    const currentStep = this.currentStep;
    if (!currentStep || currentStep.durationType !== DurationType.DISTANCE) {
      return null;
    }
    const currentDistance = this.metricService.getByMetricType(MetricType.DISTANCE)?.getLastValue() || 0;
    return currentStep.durationValue - (currentDistance - this.currentStepStartDistance);
  }

  get targetMetric(): Metric | null {
    const currentStep = this.currentStep;
    if (!currentStep) {
      return null;
    }
    switch (currentStep.targetType) {
      case TargetType.POWER:
        return this.metricService.getByMetricType(MetricType.POWER);
      case TargetType.HEART_RATE:
        return this.metricService.getByMetricType(MetricType.HEART_RATE);
      case TargetType.SPEED:
        return this.metricService.getByMetricType(MetricType.SPEED);
      case TargetType.CADENCE:
        return this.metricService.getByMetricType(MetricType.CADENCE);
      default:
        return null;
    }
  }

}


// Example workout
// {
//   "fileIdMesgs": [
//     {
//       "manufacturer": "development",
//       "type": "workout",
//       "product": 1,
//       "productName": "Intervals.icu"
//     }
//   ],
//     "workoutMesgs": [
//       {
//         "sport": "cycling",
//         "wktName": "Test",
//         "numValidSteps": 4
//       }
//     ],
//       "workoutStepMesgs": [
//         {
//           "durationType": "distance",
//           "durationValue": 500000,
//           "targetType": "power3s",
//           "customTargetValueLow": 1173,
//           "customTargetValueHigh": 1269,
//           "targetValue": 0,
//           "wktStepName": "Text to display",
//           "intensity": "interval",
//           "messageIndex": 0,
//           "durationDistance": 5000
//         },
//         {
//           "durationType": "open",
//           "targetType": "open",
//           "wktStepName": "Press lap",
//           "intensity": "warmup",
//           "messageIndex": 1
//         },
//         {
//           "durationType": "time",
//           "durationValue": 1200000,
//           "targetType": "cadence",
//           "customTargetValueLow": 87,
//           "customTargetValueHigh": 93,
//           "targetValue": 0,
//           "wktStepName": "Rmp text",
//           "intensity": "recovery",
//           "messageIndex": 2,
//           "durationTime": 1200,
//           "customTargetCadenceLow": 87,
//           "customTargetCadenceHigh": 93,
//           "targetCadenceZone": 0
//         },
//         {
//           "durationType": "time",
//           "durationValue": 20000,
//           "targetType": "powerLap",
//           "customTargetValueLow": 1468,
//           "customTargetValueHigh": 4096,
//           "targetValue": 0,
//           "wktStepName": "Max effort please",
//           "intensity": "interval",
//           "messageIndex": 3,
//           "durationTime": 20
//         }
//       ]
// }

File: android/app/src/main/java/com/openbikecomputer/service/TrainingService.java
package com.openbikecomputer.service;

import com.openbikecomputer.fit.FitDecoder;
import com.openbikecomputer.model.Metric;
import com.openbikecomputer.model.MetricType;
import com.openbikecomputer.model.Workout;
import com.openbikecomputer.model.Workout.DurationType;
import com.openbikecomputer.model.Workout.TargetType;
import com.openbikecomputer.model.Workout.WorkoutStep;

import java.io.IOException;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

/**
 * Loads structured workouts (FIT) and drives step progression.
 * Ported from {@code training.service.ts}.
 */
public class TrainingService {

    private static final long CHECK_INTERVAL_MS = 1000;

    /** Listener notified when a workout is loaded. */
    public interface OnNewWorkout {
        void onNewWorkout(Workout workout);
    }

    /** Listener notified when a step completes automatically (e.g. to play a sound). */
    public interface OnStepCompleted {
        void onStepCompleted();
    }

    private Workout workout = null;
    private final CopyOnWriteArrayList<OnNewWorkout> newWorkoutListeners = new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<OnStepCompleted> stepCompletedListeners = new CopyOnWriteArrayList<>();

    private int currentStepIndex = -1;
    private long currentStepStartTime = 0;
    private double currentStepStartDistance = 0;

    private ScheduledExecutorService scheduler;
    private ScheduledFuture<?> checkTask;

    private final MetricService metricService;

    public TrainingService(MetricService metricService) {
        this.metricService = metricService;
    }

    public void loadFitFile(byte[] fileContent, double powerThreshold, double heartRateThreshold) {
        stopWorkout();
        try {
            this.workout = FitDecoder.decodeWorkout(fileContent, powerThreshold, heartRateThreshold);
        } catch (IOException e) {
            throw new RuntimeException("Failed to read FIT file", e);
        }
        for (OnNewWorkout listener : newWorkoutListeners) {
            listener.onNewWorkout(this.workout);
        }
    }

    public void addNewWorkoutListener(OnNewWorkout listener) {
        newWorkoutListeners.add(listener);
    }

    public void removeNewWorkoutListener(OnNewWorkout listener) {
        newWorkoutListeners.remove(listener);
    }

    public void addStepCompletedListener(OnStepCompleted listener) {
        stepCompletedListeners.add(listener);
    }

    public void removeStepCompletedListener(OnStepCompleted listener) {
        stepCompletedListeners.remove(listener);
    }

    public Workout getCurrentWorkout() {
        return workout;
    }

    public boolean isRunning() {
        return currentStepIndex != -1 && workout != null;
    }

    public void startWorkout() {
        if (workout == null) {
            throw new IllegalStateException("No workout loaded");
        }
        if (currentStepIndex != -1) {
            throw new IllegalStateException("Workout already started");
        }
        startCheckInterval();
        moveToNextStep();
    }

    public void stopWorkout() {
        stopCheckInterval();
        workout = null;
        currentStepIndex = -1;
        currentStepStartTime = 0;
        currentStepStartDistance = 0;
    }

    private void startCheckInterval() {
        if (checkTask != null) {
            return;
        }
        scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "training-check");
            t.setDaemon(true);
            return t;
        });
        checkTask = scheduler.scheduleAtFixedRate(this::checkWorkoutUpdate,
                CHECK_INTERVAL_MS, CHECK_INTERVAL_MS, TimeUnit.MILLISECONDS);
    }

    private void stopCheckInterval() {
        if (checkTask != null) {
            checkTask.cancel(false);
            checkTask = null;
        }
        if (scheduler != null) {
            scheduler.shutdownNow();
            scheduler = null;
        }
    }

    protected void checkWorkoutUpdate() {
        if (workout == null || currentStepIndex < 0) {
            return;
        }
        WorkoutStep currentStep = getCurrentStep();
        if (currentStep == null) {
            return;
        }
        Double remainingTime = getRemainingTime();
        if (remainingTime != null && remainingTime <= 0) {
            moveToNextStep();
            notifyStepCompleted();
            return;
        }
        Double remainingDistance = getRemainingDistance();
        if (remainingDistance != null && remainingDistance <= 0) {
            moveToNextStep();
            notifyStepCompleted();
        }
    }

    private void notifyStepCompleted() {
        for (OnStepCompleted listener : stepCompletedListeners) {
            listener.onStepCompleted();
        }
    }

    public void moveToNextStep() {
        if (workout == null) {
            throw new IllegalStateException("No workout loaded");
        }
        currentStepIndex++;
        currentStepStartTime = System.currentTimeMillis();
        currentStepStartDistance = lastDistance();
        metricService.newLap();
    }

    public void moveToPreviousStep() {
        if (workout == null || currentStepIndex <= 0) {
            throw new IllegalStateException("No previous step available");
        }
        currentStepIndex--;
        currentStepStartTime = System.currentTimeMillis();
        currentStepStartDistance = lastDistance();
        metricService.newLap();
    }

    public WorkoutStep getCurrentStep() {
        if (workout == null || currentStepIndex >= workout.steps.size() || currentStepIndex < 0) {
            return null;
        }
        return workout.steps.get(currentStepIndex);
    }

    public WorkoutStep getNextStep() {
        if (workout == null || currentStepIndex + 1 >= workout.steps.size()) {
            return null;
        }
        return workout.steps.get(currentStepIndex + 1);
    }

    public Double getRemainingTime() {
        if (workout == null || currentStepIndex < 0) {
            return null;
        }
        WorkoutStep currentStep = getCurrentStep();
        if (currentStep == null || currentStep.durationType != DurationType.TIME) {
            return null;
        }
        double elapsedTime = (System.currentTimeMillis() - currentStepStartTime) / 1000.0;
        return currentStep.durationValue - elapsedTime;
    }

    public Double getRemainingDistance() {
        if (workout == null || currentStepIndex < 0) {
            return null;
        }
        WorkoutStep currentStep = getCurrentStep();
        if (currentStep == null || currentStep.durationType != DurationType.DISTANCE) {
            return null;
        }
        double currentDistance = lastDistance();
        return currentStep.durationValue - (currentDistance - currentStepStartDistance);
    }

    public Metric getTargetMetric() {
        WorkoutStep currentStep = getCurrentStep();
        if (currentStep == null) {
            return null;
        }
        switch (currentStep.targetType) {
            case POWER:
                return metricService.getByMetricType(MetricType.POWER);
            case HEART_RATE:
                return metricService.getByMetricType(MetricType.HEART_RATE);
            case SPEED:
                return metricService.getByMetricType(MetricType.SPEED);
            case CADENCE:
                return metricService.getByMetricType(MetricType.CADENCE);
            default:
                return null;
        }
    }

    public boolean hasTraining() {
        return workout != null;
    }

    public String getTrainingName() {
        return workout != null ? workout.name : "";
    }

    private double lastDistance() {
        Double v = metricService.getByMetricType(MetricType.DISTANCE).getLastValue();
        return v != null ? v : 0.0;
    }
}
